test/test_linearizer.py

```diff
@@ -191,14 +186,95 @@ def test_early_end_local(self):
+    k = Linearizer(*ast)
+    k.hand_coded_optimizations()
+    k.linearize()
+    local_buf = [u for u in k.uops if u.uop is UOps.DEFINE_LOCAL]
+    self.assertEqual(len(real_local_stores:=[u for u in k.uops if u.uop is UOps.STORE and any([lb in u.vin for lb in local_buf])]), 3, \
+      f"should have generated 3 BufferOps.STORE to the local buf but got {len(real_local_stores)}")
+    self.assertEqual(len(real_local_loads:=[u for u in k.uops if u.uop is UOps.LOAD and any([lb in u.vin for lb in local_buf])]), 3, \
+      f"should have generated 3 BufferOps.LOAD to the local buf but got {len(real_local_loads)}")
+    self.assertEqual((real_local_stores[1].vin[1].uop, real_local_stores[1].vin[1].arg), (UOps.CONST, 0))
+    self.assertEqual((real_local_loads[1].vin[1].uop, real_local_loads[1].vin[1].arg), (UOps.CONST, 0))
+    x = Tensor.randn(3,27,32).realize()
+    helper_linearizer_ast(ast, [x], wanna_output=[x.numpy().std(axis=2, ddof=0).reshape(-1)])
+
+  @unittest.skip
+  def test_multireduce_upcasting(self):
+    ast = LazyOp(op=BufferOps.STORE, src=(LazyOp(op=ReduceOps.SUM, src=(LazyOp(op=BinaryOps.SUB, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=MemBuffer(idx=1, dtype=dtypes.float32, st=ShapeTracker(views=(View(shape=(2, 7), strides=(7, 1), offset=0, mask=None, contiguous=True),)))), LazyOp(op=ReduceOps.SUM, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=MemBuffer(idx=1, dtype=dtypes.float32, st=ShapeTracker(views=(View(shape=(2, 7), strides=(7, 1), offset=0, mask=None, contiguous=True),),))),), arg=(1,)),)),), arg=(1,)),), arg=MemBuffer(idx=0, dtype=dtypes.float32, st=ShapeTracker(views=(View(shape=(2, 1), strides=(1, 0), offset=0, mask=None, contiguous=True),)))), # noqa: E501
+    k = Linearizer(*ast)
+    k.upcast()
+    k.linearize()
+    define_globals = [u for u in k.uops.uops if u.uop is UOps.DEFINE_GLOBAL]
+    self.assertEqual(len([u for u in k.uops.uops if u.uop is UOps.LOAD and define_globals[1] in u.vin]), 7)
+    self.assertEqual(len([u for u in k.uops.uops if u.uop is UOps.ALU and u.arg is BinaryOps.SUB]), 7)
+    x = Tensor.randn(2,7).softmax().realize()
+    helper_linearizer_ast(ast, [x], wanna_output=[(x.numpy() - x.numpy().sum(axis=1, keepdims=True)).sum(axis=1)])
+
+  @unittest.skip
+  def test_multireduce_loop_scope(self):
```

Comment:
with multiple reductions, operations on the result of one reduction may get generated while rendering another so the ALU uops would get placed within the loop. This example calculates a layernorm of a tensor then sends it through another `SUM` to illustrate the issue

```
  float acc0 = 0.0f;
  for (int ridx0 = 0; ridx0 < 32; ridx0++) {
    float val0 = data1[alu0+ridx0];
    acc0 = (val0+acc0);
  }
+ float alu1 = (acc0*0.03125f); // we want the mean calculation to go here, so it's in all scopes
  float acc1 = 0.0f;
  for (int ridx1 = 0; ridx1 < 32; ridx1++) {
    float val1 = data1[alu0+ridx1];
-   float alu2 = (val1-acc0*0.03125f); // the linearizer generates the mean here while rendering the std reduction
+   float alu2 = (val1-alu1); 
    acc1 = ((alu2*alu2)+acc1);
  }
  float acc2 = 0.0f;
  for (int ridx2 = 0; ridx2 < 32; ridx2++) {
    float val2 = data1[alu0+ridx2];
    acc2 = (((val2-alu1)/sqrt((acc1*0.03125f)))+acc2); // BUT the kernel also uses the mean here, and unchanged it will be out-of-scope
  }
  data0[gidx0] = acc2;
```